### PR TITLE
Added pods-per-core to kubelet. #25762

### PR DESCRIPTION
**Pull request by rrati**
Added --pods-per-core to kubelet
#25762

### CODE REVIEW

**Comment by timothysc** [comment]
@rrati needs a test.

File: cmd/kubelet/app/options/options.go [hunk 1]
```diff
@@ -258,4 +259,5 @@ func (s *KubeletServer) AddFlags(fs *pflag.FlagSet) {
 	fs.StringVar(&s.EvictionSoftGracePeriod, "eviction-soft-grace-period", s.EvictionSoftGracePeriod, "A set of eviction grace periods (e.g. memory.available=1m30s) that correspond to how long a soft eviction threshold must hold before triggering a pod eviction.")
 	fs.DurationVar(&s.EvictionPressureTransitionPeriod.Duration, "eviction-pressure-transition-period", s.EvictionPressureTransitionPeriod.Duration, "Duration for which the kubelet has to wait before transitioning out of an eviction pressure condition.")
 	fs.Int32Var(&s.EvictionMaxPodGracePeriod, "eviction-max-pod-grace-period", s.EvictionMaxPodGracePeriod, "Maximum allowed grace period (in seconds) to use when terminating pods in response to a soft eviction threshold being met.  If negative, defer to pod specified value.")
+	fs.Int32Var(&s.PodsPerCore, "pods-per-core", s.PodsPerCore, "Number of Pods per core that can run on this Kubelet. A value of 0 disables this limit. Cannot exceed max-pods")
```
Comment (wojtek-t) [inline cmd/kubelet/app/options/options.go]: What happens if it will exceed max-pod?

Also, this cannot exeed max-pod or this * number of cores?

**Comment by wojtek-t** [comment]
lgtm - just one minor comment

File: cmd/kubelet/app/options/options.go (hunk 1 above)
Comment (rrati) [inline cmd/kubelet/app/options/options.go]: max-pods is the total # of pods allowed, so if pods per core * cores > max-pods, then the kubelet will report max-pods.  Should I make the statement more clear?

Comment (wojtek-t) [inline cmd/kubelet/app/options/options.go]: yes please

**Comment by rrati** [comment]
tests with 2 core system:

kubelet --pods-per-core=20
kubectl describe node
    Capacity:
        pods:				40
        cpu:					2

kubelet --pods-per-core=100
kubectl describe node
    Capacity:
        pods:				110
        cpu:					2

kubelet --pods-per-core=10 --max-pods=10
kubectl describe node
    Capacity:
        pods:				10
        cpu:					2

**Comment by rrati** [comment]
Logged issue #25875 to address automated testing

**Comment by wojtek-t** [comment]
lgtm

**Comment by rrati** [comment]
@k8s-bot Please unit test this issue #IGNORE

**Comment by rrati** [comment]
@k8s-bot unit test this issue #IGNORE

**Comment by yujuhong** [comment]
Do we plan to use this flag for v1.3?

/cc @dchen1107

**Comment by vishh** [comment]
@yujuhong https://github.com/kubernetes/kubernetes/issues/25762#issuecomment-220043356

**Comment by wojtek-t** [comment]
@yujuhong no, it will be 0 by default; it was just requested by RedHat. I would like to take advantage of it in the next release hopefully.

**Comment by rrati** [comment]
@k8s-bot test this issue #IGNORE

**Comment by jeremyeder** [comment]
For docs:

"pods-per-core" is a simple scaling factor that allows us to get past the previous limit of 110 on larger systems.

OpenShift could lay down a config like so (via installer):

--pods-per-core=10 
--max-pods=255

Which, when compared with openshift 3.2 max-pods=110, the effective behavior of the new pods-per-core algorithm:

For 1-10 cores, max-pods is reduced
For 11+ cores, max-pods is increased

Examples:

8 cores:
8 cores * pods-per-core(10) = 80
80 < max-pods, so the limit is 80 pods.

16 cores:
16 cores * pods-per-core(10) = 160
160 < max-pods, so the limit is 160 pods.

32 cores * pods-per-core(10) = 320
320 > max-pods, so max-pods wins and the limit is 255

**Comment by alex-mohr** [comment]
Buildcop: Seems worth a release-note, so swapped labels.  LMK if that's wrong.

**Comment by jeremyeder** [comment]
Should the code throw an error if running_pods > max_pods ?

**Comment by yujuhong** [comment]
> Should the code throw an error if running_pods > max_pods ?

That's a valid question. I think after a restart, kubelet will admit each pod again, and some of the pods will get rejected due the MaxPods restriction. The containers belonging to those pods will be stopped.

**Comment by warmchang** [comment]
mark.

**Pull request merged**